### PR TITLE
nixos/nextcloud-notify_push: fix connecting to mysql via socket

### LINKED ISSUE #310335: nextcloud28 notify_push Failed to connect to database
### Describe the bug
I installed Nextcloud on a fresh NixOS installation. Nextcloud itself works flawlessly, but the notify_push service does not start.

journalctl -xeu nextcloud-notify_push.service

```
nextcloud-notify_push-start[762819]: Error:   × Failed to connect to database
nextcloud-notify_push-start[762819]:   ├─▶ error returned from database: 1698 (28000): Access denied for user
nextcloud-notify_push-start[762819]:   │   'nextcloud'@'localhost'
nextcloud-notify_push-start[762819]:   ╰─▶ 1698 (28000): Access denied for user 'nextcloud'@'localhost'
systemd[1]: nextcloud-notify_push.service: Main process exited, code=exited, status=1/FAILURE
...
...
nextcloud-notify_push-post-start[762818]: ✓ redis is configured
nextcloud-notify_push-post-start[762818]: 🗴 can't connect to push server: Server error: `GET https://cloud.****.com/push/test/cookie` resulted in a `502 Bad Gateway` response:
nextcloud-notify_push-post-start[762818]: <html>
nextcloud-notify_push-post-start[762818]: <head><title>502 Bad Gateway</title></head>
nextcloud-notify_push-post-start[762818]: <body>
nextcloud-notify_push-post-start[762818]: <center><h1>502 Bad Gateway</h1></center>
nextcloud-notify_push-post-start[762818]: <hr><center>ngin (truncated...)
May 09 14:38:29 stelzen-art systemd[1]: nextcloud-notify_push.service: Control process exited, code=exited, status=1/FAILURE
```

```
MariaDB [(none)]> SHOW GRANTS FOR 'nextcloud'@'localhost';
+--------------------------------------------------------------------------+
| Grants for nextcloud@localhost                                           |
+--------------------------------------------------------------------------+
| GRANT USAGE ON *.* TO `nextcloud`@`localhost` IDENTIFIED VIA unix_socket |
| GRANT ALL PRIVILEGES ON `nextcloud`.* TO `nextcloud`@`localhost`         |
+--------------------------------------------------------------------------+
```

### Steps To Reproduce
Config:
```
services.nextcloud = {
    enable = true;
    package = pkgs.nextcloud29;
    hostName = "cloud.*******.com";
    database.createLocally = true;
    config = {
      dbtype = "mysql";
      adminuser = "admin";
      adminpassFile = "/etc/nextcloud-admin-pass";
    };
    
    extraOptions = {
      allow_user_to_change_display_name = true;
      maintenance_window_start = 3;
    };
    
    https = true;
		 
    caching = {
        apcu = true;
    }; 
    
    notify_push = {
    	enable = true;
    };
    
    appstoreEnable = true;
    extraAppsEnable = true;
    extraApps = { inherit (pkgs.nextcloud28Packages.apps) "notify_push"; };
};

services.nginx.virtualHosts.${config.services.nextcloud.hostName} = {
    forceSSL = true;
    enableACME = true;
};
```

```
services.mysql = {
    enable = true;
    package = pkgs.mariadb;
    dataDir = "/var/lib/mysql";
};
```

### Expected behavior
nextcloud-notify_push works

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.30, NixOS, 23.11 (Tapir), 23.11.6981.27c13997bf45`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `"nixos-23.11-small"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

@schneefux @bachp @globin @Ma27

### PR DESCRIPTION
**Pull request by st3iny**
Fix https://github.com/NixOS/nixpkgs/issues/310335

Hey folks, I'm a Nextcloud developer and am currently tinkering with NixOS. Here is a little fix I came up with.

Instructing Nextcloud to connect to MySQL via a socket requires setting dbhost to `localhost:/run/mysqld/mysqld.sock` while Postgresql requires just the path to the socket.

Thus, the notify_push service wasn't able to connect to MySQL via a socket due to a malformed `DATABASE_URL` env variable. The host part is omitted if a socket is detected.

I refactored and fixed the logic when MySQL is used to always include the host part and also add the socket option to the `DATABASE_URL` variable. Notify_push will now connects to the MySQL socket properly.

I tested this locally with a fresh Postgres and MySQL database to make sure that there are no regressions.

```sh
# Generated strings:
DATABASE_URL='mysql://nextcloud@localhost/nextcloud?socket=/run/mysqld/mysqld.sock'
DATABASE_URL='postgresql://nextcloud/nextcloud?host=/run/postgresql'
```

This is my first PR here so please let me know in case something needs to be adjusted.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by st3iny** [comment]
cc @SuperSandro2000

**Comment by SuperSandro2000** [comment]
Thanks! I don't use mysql/mariadb so I must admit I myself never tested this. 😅

File: nixos/modules/services/web-apps/nextcloud-notify_push.nix [hunk 1]
```diff
@@ -77,15 +77,22 @@ in
...
+        dbHostHasPrefix = prefix: lib.hasPrefix prefix (toString cfg.dbhost);
+        isPostgresql = dbType == "postgresql";
+        isMysql = dbType == "mysql";
+        isSocket = (isPostgresql && dbHostHasPrefix "/") || (isMysql && dbHostHasPrefix "localhost:/");
         dbHost = lib.optionalString (cfg.dbhost != null) (if
           isSocket then
-            if dbType == "postgresql" then "?host=${cfg.dbhost}" else
-            if dbType == "mysql" then "?socket=${cfg.dbhost}" else throw "unsupported dbtype"
+            if isPostgresql then "" else
+            if isMysql then "@localhost" else throw "unsupported dbtype"
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/nextcloud-notify_push.nix:87]: We don't really want to throw inside the module system as that will not show all errors/warnings but that was already here before and we would need to refactor the database configuration a lot to bring that up to the right level. 😅

Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/nextcloud-notify_push.nix:87]:
```suggestion
            lib.optionalString isMysql "@localhost"
```
we already throw in line 92, so we can condense this here a bit.

Comment (st3iny) [reply nixos/modules/services/web-apps/nextcloud-notify_push.nix:87]: I fully agree. This whole nested if expression also looks a bit ugly.

Comment (st3iny) [reply nixos/modules/services/web-apps/nextcloud-notify_push.nix]: Done

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by SuperSandro2000** [comment]
Tested this on my installation which uses postgres and which still works afterwards.

**Pull request merged**